Basic frontend

lightvector opened the pull request:
Implements pages for networks and games.

The CSS on them isn't working right now - the tables of networks and games are all mashed together and not spaced properly. But I think the content is reasonable. The front page and contributors page are still remaining to be handled. It would also be nice to have some summary stats (like total games, and total games by network, and things like that).

@tychota - please let me know if there is anything in this implementation that looks suspect to you, since this is my first attempt and I am only just learning how all this works. Also, in the pagination, are there going to be major performance concerns when we scale up, given how it is implemented?

tychota commented on src/frontend/templates/pages/games_list.html:
```diff
@@ -0,0 +1,152 @@
+    <a href="{% url 'training_games_list' run.name network.name 1 %}">&laquo; First</a>
+    <a href="{% url 'training_games_list' run.name network.name page_obj.previous_page_number %}">Prev</a>
+    {% else %}
+    &laquo; First
```
We can use font awesome icons instead of &laquo;

Eg `<i class="fab fa-github"></i>` will display github icon

tychota replied:
https://fontawesome.com/icons?d=gallery

lightvector replied:
Ooh nice icons. Okay done!

tychota replied:
nicee

tychota commented on src/frontend/templates/pages/games_list.html:
```diff
@@ -0,0 +1,152 @@
+</div>
+{% endif %}
+
+<table>
```
You can use the css for table

tychota replied:
https://bulma.io/documentation/elements/table/
![Uploading image.png…]()

lightvector replied:
Done!

tychota reviewed (commented):
Only a partial review, will finish the review later

lightvector commented on src/frontend/templates/base.html:
```diff
@@ -39,10 +39,10 @@
       <div id="navMenuExample" class="navbar-menu">
         <div class="navbar-start">
-          <a class="navbar-item is-active" href="/frontend/networks/">
+          <a class="navbar-item is-active" href="{% url 'current_run_networks' %}">
```
thanks

lightvector commented on src/apps/games/models/abstract_game.py:
```diff
@@ -130,13 +130,18 @@ class GamesResult(TextChoices):
             return f"{self.winner}+R"
-        if self.winner not in [AbstractGame.GamesResult.BLACK, AbstractGame.GamesResult.WHITE]:
-            return self.winner
+        elif self.winner == AbstractGame.GamesResult.DRAW:
```
Yes, but only to single characters "0" and "-". When I looked at how this looked on the actual website, I immediately didn't like it - there will be many users who are confused by it much more than the words "Draw" and "Void".

lightvector replied:
I'm going to leave this alone for now, I think it's fine. The default text for a `GamesResult` is suitable for a 1-character display, and whereas `result_text` is a function that produces a string that is suitable for human viewing, which means that having a little more logic to spell out things like "Draw" seems okay to me.

lightvector commented on src/apps/games/models/abstract_game.py:
```diff
@@ -130,13 +130,18 @@ class GamesResult(TextChoices):
-        else:
+        elif self.winner == AbstractGame.GamesResult.WHITE:
             score = self.score
+        else:
```
I figure that if we ever decide to add more result types, but we forget to update the code here to translate the result to text, then this will be more reasonable in the meantime before we fix the code, instead of assuming that the score should be translated as if the player were white.

lightvector commented on src/frontend/templates/base.html:
```diff
@@ -104,16 +104,17 @@
+          (<a href="http://opensource.org/licenses/mit-license.php">server code license</a>).
         </p>
+        <p><a href="/about">Discover more about the distributed training project here</a></p>
+        <!-- <p> -->
```
Oh, I thought the little Github icon link is redundant now that the earlier message has direct links to both the engine and server githubs. I guess I should just delete instead of comment, right?

lightvector replied:
Deleted, and moved the github icons to the new links. Thanks.

lightvector commented on src/frontend/templates/pages/game_network_groups.html:
```diff
@@ -0,0 +1,49 @@
+    <th>  </th>
+    <th>  </th>
+  </tr>
+  {% for network in object_list %}
+  <tr>
+    <td> {{ network.name }} </td>
+    <td> <a href="{% url 'training_games_list' run.name network.name 1 %}"> Training Games </td>
+    <td> <a href="{% url 'rating_games_list' run.name network.name 1 %}"> Rating Games </td>
+  </tr>
+  {% endfor %}
```
Do you mean "inefficient" for performance, or "inefficient" for the user UI?

If you mean "inefficient" for the user UI, then I'm curious what a better design is. The design you suggests has a major problem which is that the number of games is so extremely large that just having the whole thing in a list view is not very useful for the user. The user needs a way to filter the games by different criteria, and jump to different sections of the games.

I didn't know how to implement a general filtering mechanism, so I figured that grouping by networks would be a reasonable choice. A lot of possible applications would be handled by grouping by networks (e.g. the user wants to see all the games that contribute to a network's Elo, or the user wants to compare how two different networks play). Also, since network ordering loosely corresponds to chronological ordering (not completely, but very close), this means that applications that want to get recent games, or games from a certain time range, can also do so approximately.

But if you think we can have a single list page and implement some effective and much more general way of user filtering, then that would be better.

tychota replied:
> I didn't know how to implement a general filtering mechanism, so I figured that grouping by networks would be a reasonable choice. A lot of possible applications would be handled by grouping by networks (e.g. the user wants to see all the games that contribute to a network's Elo, or the user wants to compare how two different networks play). Also, since network ordering loosely corresponds to chronological ordering (not completely, but very close), this means that applications that want to get recent games, or games from a certain time range, can also do so approximately.

Indeed. We can use django filter later https://django-filter.readthedocs.io/en/master/guide/usage.html#the-view

---

On think use need is the latest games he contributed, but it will be in contribution page.

lightvector commented on src/frontend/templates/pages/game_network_groups.html:
```diff
@@ -0,0 +1,49 @@
+<h1 class="title">Games for {{run.name}} by Network</h1>
+{% endif %}
+
+{% if run and object_list|length <= 0 %}
```
Done, thanks!

lightvector commented:
Added an SGF viewer as well!
http://wgo.waltheri.net/

Except that it doesn't support rectangular boards. So... we're going to either have to use a different player (which seems tricky, I bet very few support rectangle sizes even though it's part of the SGF standard), or go through and edit the source code to wgo.js to support it.

lightvector commented:
![image](https://user-images.githubusercontent.com/11942395/96324715-29567900-0ff1-11eb-8748-ac275aee5a43.png)

tychota commented on src/frontend/templates/pages/games_list.html:
```diff
@@ -0,0 +1,152 @@
+  </span>
+</div>
+{% else %}
+<div class="pagination">
```
You may want to extract this.

https://stackoverflow.com/a/52231845

tychota commented on src/frontend/views/games_view.py:
```diff
@@ -0,0 +1,75 @@
+    # return render(request, 'list.html', {'page_obj': page_obj})
+
+    # # We're returning the networks, because we display games organized by network
+    # return Game.objects.filter(run=self.run).order_by("-created_at")
```
Is it still useful ?

lightvector replied:
Deleted it.

tychota commented on src/frontend/templates/pages/game_network_groups.html:
```diff
@@ -21,11 +21,11 @@ <h1 class="title">Games for {{run.name}} by Network</h1>
 No games found for this run!
 {% elif run and object_list|length > 0 %}
-<table>
+<table class="table">
```
Nice

tychota commented on src/frontend/templates/pages/sgfplayer.html:
```diff
@@ -0,0 +1,42 @@
+<li> Komi: {{ game.komi }} </li>
+<li> Game Length: {{ game.game_length }} moves </li>
+<li> Result: {{ game.result_text }} </li>
+</ul>
```
I think a table will looks better, maybe in a box (https://bulma.io/documentation/elements/box/)

![image](https://user-images.githubusercontent.com/13785185/96377531-7c156980-1186-11eb-920c-c5107ff17f17.png)

lightvector replied:
I went ahead and stuck the whole thing in a content box, but maybe we can do a table later.

tychota commented on src/frontend/templates/pages/sgfplayer.html:
```diff
@@ -0,0 +1,42 @@
+{% extends 'base.html' %}
```
Nice

tychota commented on src/frontend/templates/pages/sgfplayer.html:
```diff
@@ -0,0 +1,42 @@
+<h1 class="title">Training Game ID {{game.id}} (run {{game.run.name}}) </h1>
+{% endif %}
+
+<script type="text/javascript" src="/static/js/wgo.min.js"></script>
+<script type="text/javascript" src="/static/js/wgo.player.min.js"></script>
+<link type="text/css" href="/static/css/wgo.player.css" rel="stylesheet" />
+<link type="text/css" href="/static/css/wgo-custom.css" rel="stylesheet" />
```
Ideally we put this a in a a `{% block X %}` (X being maybe "Head", not sure if the is one think that exists like this right now) so the js file are in the head section of HTML and not the body, like other js and css.

lightvector replied:
Done.

tychota commented:
Also It is weird to see pagination if there is no game or even a single page

![image](https://user-images.githubusercontent.com/13785185/96377505-630cb880-1186-11eb-930b-9df4510fed85.png)

lightvector commented on src/frontend/templates/pages/networks.html:
```diff
@@ -1,57 +1,61 @@
 
-{% block title %}Bulma Site{% endblock %}
+{% block title %}
+{% if not run %}
```
Thanks. My emacs doesn't auto-indent because it recognizes html but it doesn't recognize django template language. I just go with the indention my emacs puts in automatically because it is easier (and doing differently requires me to fight the editor, which is not fun).

I think the block structure is hopefully not too bad because I use additional blank lines in places to try to give visual separation when appropriate.

lightvector commented on src/frontend/templates/pages/networks.html:
```diff
@@ -1,57 +1,61 @@
+{% elif run and object_list|length > 0 %}
+<table>
+  <tr>
+    <th> Name </th>
```
It doesn't do anything right?

lightvector replied:
I consider this no harm, just up to the code formatting. There are sometimes spaces or not elsewhere too.

lightvector commented on src/frontend/templates/pages/networks.html:
```diff
@@ -1,57 +1,61 @@
+  <li><a href="{% url 'networks' run.name %}"> {{ run.name }}</a></li>
+  {% endfor %}
+</ul>
+{% endif %}
```
This is used when:

* There is more than one run. In that case, this provides a table of links to the networks for older runs besides the current run.

OR

* There is no current run because none of the runs are active right now (even if there is only 1 run). In that case, this provides a table of links to the networks for all runs.

lightvector commented on src/frontend/templates/pages/networks.html:
```diff
@@ -1,57 +1,61 @@
-{% block title %}Bulma Site{% endblock %}
+{% block title %}
+{% if not run %}
+Networks
```
Fixed! :)

lightvector commented on src/frontend/templates/pages/networks.html:
```diff
@@ -1,57 +1,61 @@
+{% endif %}
+
+{% if run and object_list|length <= 0 %}
+No networks found for this run!
```
`I think it's cleaner to have fewer variables and less action-at-a-distance in the code. 
The `object_list` (now renamed to `networks_list`) networks list is exactly a variable that is looped over just below in the same template. So it makes sense to have the check on that variable here, instead of having to know that `run_has_no_network` is equivalent to `networks_list` being length 0.

lightvector commented on src/frontend/views/networks_view.py:
```diff
@@ -0,0 +1,38 @@
+      )
+      if show_all_runs:
+        context["all_runs_to_show"] = list(Run.objects.all().order_by("-created_at"))
+      return context
```
No, this logic should not be eliminated. It is the logic that implements the question you asked `When is this used ?` in the other comment.

lightvector commented on src/frontend/views/networks_view.py:
```diff
@@ -0,0 +1,38 @@
+
+    if self.run is None:
+      return []
+    return Network.objects.filter(run=self.run).order_by("-created_at")
```
There is never a column "run" because there is only ever one active run at a time. But I did simplify the logic here, thanks!

lightvector commented on src/urls.py:
```diff
@@ -67,15 +69,16 @@
-        path("frontend/networks/", TemplateView.as_view(template_name="pages/networks.html"), name="networks", ),
-        path("frontend/games/", TemplateView.as_view(template_name="pages/games.html"), name="games", ),
-        path("frontend/contributions/", TemplateView.as_view(template_name="pages/contributions.html"), name="contributions", ),
+        path("frontend/networks/", NetworksView.as_view(template_name="pages/networks.html"), {"viewing_current_run": True}, name="current_run_networks"),
```
Thanks, done.

lightvector commented:
Okay I did all the translations and addressed most of your comments. I also fixed up and improved a lot of the details of formatting and stuff. Will do more work later on adding the front page and contributors page.

lightvector commented:
And if you would like to take a quick look at the latest rounds of commits, that would be cool, if you want to until I push some of the next new pages to review further, that would be cool too.

Thanks for the help!

tychota commented on src/frontend/static/js/wgo.min.js:
```diff
@@ -1,51 +1,76 @@
-b)&&0!=this.position.get(a,b)?(this.position.set(a,b,0),!0):!1},setStone:function(a,b,c){return this.isOnBoard(a,b)?(this.position.set(a,b,c||0),!0):!1},getStone:function(a,b){return this.isOnBoard(a,b)?this.position.get(a,b):0},pushPosition:function(a){a||(a=this.position.clone(),a.capCount={black:this.position.capCount.black,white:this.position.capCount.white},a.color=this.position.color);this. ...
-(a=this.stack.pop(),this.turn=0==this.stack.length?g.B:this.position.color?-this.position.color:-this.turn);return a},firstPosition:function(){this.stack=[];this.stack[0]=new s(this.size);this.stack[0].capCount={black:0,white:0};this.turn=g.B;return this},getCaptureCount:function(a){return a==g.B?this.position.capCount.black:this.position.capCount.white},validatePosition:function(){for(var a,b,c= ...
-k))b=d.length,d=d.concat(l(f,h-1,k,-a),l(f,h+1,k,-a),l(f,h,k-1,-a),l(f,h,k+1,-a)),a==g.B?e+=d-b:c+=d-b;this.position.capCount.black+=e;this.position.capCount.white+=c;this.position.schema=f.schema;return d}};g.Game=m;v.WGo=g})(window);
+/*! MIT license, more info: wgo.waltheri.net *//*
```
Lol what did change ?

tychota replied:
Ok, Ive seen https://github.com/waltheri/wgo.js/pull/105/files

lightvector replied:
Yep. :)

tychota commented on src/frontend/static/bulma/css/style.css:
```diff
@@ -374,7 +377,7 @@ a:hover {
 code {
   background-color: whitesmoke;
-  color: #f14668;
+  color: #da1039;
```
Out of curiosity, what drived this change / upgrade of bulma ?

lightvector replied:
I wanted to use some helper css classes in bulma ("m-*") but I discovered they weren't there yet in our version.
